engine/src/main/java/org/pentaho/di/trans/steps/validator/ValidatorMeta.java

```diff
@@ -0,0 +1,351 @@
+import org.pentaho.di.trans.step.errorhandling.Stream;
+import org.pentaho.di.trans.step.errorhandling.StreamIcon;
+import org.pentaho.di.trans.step.errorhandling.StreamInterface;
+import org.pentaho.di.trans.step.errorhandling.StreamInterface.StreamType;
+import org.pentaho.metastore.api.IMetaStore;
+import org.w3c.dom.Node;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Contains the meta-data for the Validator step: calculates predefined formula's
+ *
+ * Created on 08-sep-2005
+ */
+@InjectionSupported( localizationPrefix = "Validator.Injection.", groups = { "VALIDATIONS" } )
+public class ValidatorMeta extends BaseStepMeta implements StepMetaInterface {
+  private static Class<?> PKG = ValidatorMeta.class; // for i18n purposes, needed by Translator2!!
+
+  /** The calculations to be performed */
+  @InjectionDeep
+  private List<Validation> validations;
+
+  /** Checkbox to have all rules validated, with all the errors in the output */
+  @Injection( name = "VALIDATE_ALL" )
+  private boolean validatingAll;
+
+  /**
+   * If enabled, it concatenates all encountered errors with the selected separator
+   */
+  @Injection( name = "CONCATENATE_ERRORS" )
+  private boolean concatenatingErrors;
+
+  /**
+   * The concatenation separator
+   */
+  @Injection( name = "CONCATENATION_SEPARATOR" )
+  private String concatenationSeparator;
+
+  public void allocate( int nrValidations ) {
+    validations = new ArrayList<Validation>( nrValidations );
+  }
+
+  public void loadXML( Node stepnode, List<DatabaseMeta> databases, IMetaStore metaStore ) throws KettleXMLException {
+    int nrCalcs = XMLHandler.countNodes( stepnode, Validation.XML_TAG );
+    allocate( nrCalcs );
+    validatingAll = "Y".equalsIgnoreCase( XMLHandler.getTagValue( stepnode, "validate_all" ) );
+    concatenatingErrors = "Y".equalsIgnoreCase( XMLHandler.getTagValue( stepnode, "concat_errors" ) );
+    concatenationSeparator = XMLHandler.getTagValue( stepnode, "concat_separator" );
+
+    for ( int i = 0; i < nrCalcs; i++ ) {
+      Node calcnode = XMLHandler.getSubNodeByNr( stepnode, Validation.XML_TAG, i );
+      validations.add( new Validation( calcnode ) );
+    }
+  }
+
+  public String getXML() {
+    StringBuilder retval = new StringBuilder( 300 );
+
+    retval.append( XMLHandler.addTagValue( "validate_all", validatingAll ) );
+    retval.append( XMLHandler.addTagValue( "concat_errors", concatenatingErrors ) );
+    retval.append( XMLHandler.addTagValue( "concat_separator", concatenationSeparator ) );
+
+    for ( int i = 0; i < validations.size(); i++ ) {
+      retval.append( "       " ).append( validations.get( i ).getXML() ).append( Const.CR );
+    }
+
+    return retval.toString();
+  }
+
+  public boolean equals( Object obj ) {
+    if ( obj != null && ( obj.getClass().equals( this.getClass() ) ) ) {
+      ValidatorMeta m = (ValidatorMeta) obj;
+      return ( getXML() == m.getXML() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.